glnx-lockfile.c

```diff
@@ -0,0 +1,192 @@
+         * are running on an older kernel, then we fall back to good
+         * old BSD locks. They also have nice semantics, but are
+         * slightly problematic on NFS, where they are upgraded to
+         * POSIX locks, even though locally they are orthogonal to
+         * POSIX locks.
+         */
+
+        t = g_strdup(p);
+
+        for (;;) {
+#ifdef F_OFD_SETLK
+                struct flock fl = {
+                        .l_type = (operation & ~LOCK_NB) == LOCK_EX ? F_WRLCK : F_RDLCK,
+                        .l_whence = SEEK_SET,
+                };
+#endif
+                struct stat st;
+
+                fd = openat(dfd, p, O_CREAT|O_RDWR|O_NOFOLLOW|O_CLOEXEC|O_NOCTTY, 0600);
+                if (fd < 0) {
+                        glnx_set_error_from_errno(error);
+                        goto out;
+                }
+
+                /* Unfortunately, new locks are not in RHEL 7.1 glibc */
+#ifdef F_OFD_SETLK
+                r = fcntl(fd, (operation & LOCK_NB) ? F_OFD_SETLK : F_OFD_SETLKW, &fl);
+#else
+                r = -1;
+                errno = EINVAL;
+#endif
+                if (r < 0) {
+
+                        /* If the kernel is too old, use good old BSD locks */
+                        if (errno == EINVAL)
+                                r = flock(fd, operation);
+
+                        if (r < 0) {
+                                glnx_set_error_from_errno(error);
+                                goto out;
+                        }
+                }
+
+                /* If we acquired the lock, let's check if the file
+                 * still exists in the file system. If not, then the
+                 * previous exclusive owner removed it and then closed
+                 * it. In such a case our acquired lock is worthless,
+                 * hence try again. */
+
+                r = fstat(fd, &st);
+                if (r < 0) {
+                        glnx_set_error_from_errno(error);
+                        goto out;
+                }
+                if (st.st_nlink > 0)
+                        break;
+
+                (void) close(fd);
+                fd = -1;
+        }
+
+        /* Note that if this is not AT_FDCWD, the caller takes responsibility
+         * for the fd's lifetime being >= that of the lock.
+         */
+        out_lock->dfd = dfd;
+        out_lock->path = t;
+        out_lock->fd = fd;
+        out_lock->operation = operation;
+
+        fd = -1;
+        t = NULL;
+
+        ret = TRUE;
+ out:
+        return ret;
+}
+
+gboolean glnx_make_lock_file_for(int dfd, const char *p, int operation, GLnxLockFile *ret, GError **error) {
+        const char *fn;
+        char *t;
+
+        g_return_val_if_fail(p != NULL, FALSE);
+        g_return_val_if_fail(ret != NULL, FALSE);
+        fn = glnx_basename(p);
+
+        t = newa(char, strlen(p) + 2 + 4 + 1);
+        stpcpy(stpcpy(stpcpy(mempcpy(t, p, fn - p), ".#"), fn), ".lck");
```

Comment:
Can you add a comment explaining how this function is different from `glnx_make_lock_file()`?  Because this line is rather cryptic.